Решение

tiamato opened the pull request:
**Задача:**
Избавьтесь от дублирующегося кода и выведите объектную модель.

**Решение:**
Выделены 3 объекта: 
* **Coordinate** - координата - реализован метод перемещения на шаг случайной величины, параметры генерации шага прибиты гвоздем внутри, т.к. у игрового объекта нет специфики, поэтому не стал выносить
* **GameObject** - игровой объект - реализованы методы убивания себя, проверка на столкновение с другим объектом и вывода в консоль:
_Kill_ - убить (себя) - не _Die_, потому что легче соотнести с именем метода по убийству столкнувшихся объектов в GameObjectList
* **GameObjectList** - список игровых объектов - реализованы методы добавления, манипуляций с игровыми объектами

Commit:
Решение

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,85 +3,127 @@
 
-                if (obj3x < 0)
-                    obj3x = 0;
+    public class GameObjectList
```
Это не просто список, у него есть конкретные ответственности далеко за рамками обычного хранения

tiamato replied:
Дополнительно выделены следующие объекты:
- **Scene** - модель, реализующая хранилище игровых объектов и логику задачи
- **ConsoleView** - представление, реализующее отображение игровых объектов
- **SceneController** - контроллер, реализующий коммуникацию представления с моделью

**GameObjectList** исключен из реализации

Commit:
Выделил модель Scene, контроллер SceneController и представление ConsoleView из GameObjectList

Commit:
Удален неиспользуемый метод GameObject.IsCollisionWith

Commit:
Удален неиспользуемый метод Coordinate.Equals

Commit:
изменен EventHandler --> Action, т.к. параметры не передаем

Commit:
Убрана инициализация _random из конструктора Scene

Commit:
Переименован метод ConsoleView.SceneUpdated --> ConsoleView.PrintAliveGameObjects

Commit:
Диапазон изменения координат StepRange вынесен в конструктор GameObject

Commit:
Исправлена ошибка при вычислении столкнувшихся объектов

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,196 @@
+        }
+    }
+
+    public interface IScene
```
Это тут зачем?

tiamato replied:
Убрал. Согласен с тем, что гибкость нужна тогда, когда она необходима, другими словами - не делаем функциональность "на будущее"

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,196 @@
+
+        private void KillGameObjectsWithCollision()
+        {
+            var itemsWithCollision = _items
+                .GroupBy(gameObject => new { X = gameObject.XCoordinate.Value, Y = gameObject.YCoordinate.Value })
+                .Where(grouping => grouping.Count() > 1)
+                .SelectMany(grouping => grouping);
+
```
Linq тут не уместно

tiamato replied:
Убрал. Руки чесались избавиться от N^2 - сделал по-другому

Commit:
Метод Scene.KillGameObjectsWithCollision реализован без использования Linq
Добавлен метод CompareTo в Coordinate
Добавлена реализация IComparable в GameObject (используется сортировка)
Удален интерфейс IScene

Commit:
Исправление ошибок оформлению кода

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
+        }
 
-            while (true)
+        private static Scene InitScene()
```
CreateScene

tiamato replied:
Исправлено

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
+            Value = coordinate;
+        }
+
+        public void RandomStep(Random random, int stepMinValue, int stepMaxValue)
```
Это не ответственность координаты. Посмотри сколько пришлось сюда передавать

tiamato replied:
Перенес в класс GameObject

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
-                obj3y += random.Next(-1, 1);
+    public class GameObject : IComparable
+    {
+        public readonly Coordinate XCoordinate;
+        public readonly Coordinate YCoordinate;
```
Посмотри в сторону типа Vector2

tiamato replied:
Координаты заменил на вектор

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
+        public readonly Coordinate XCoordinate;
+        public readonly Coordinate YCoordinate;
+        public readonly string Name;
+        public bool IsAlive { get; private set; }
```
Действительно ли ему нужен этот флаг?

tiamato replied:
нужен. Объекты из коллекции удалять нельзя, см. ниже

tiamato replied:
Флаг удален, см. ниже

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
+    {
+        public readonly Coordinate XCoordinate;
+        public readonly Coordinate YCoordinate;
+        public readonly string Name;
+        public bool IsAlive { get; private set; }
+        private readonly StepRange _stepRange;
```
https://stackoverflow.com/questions/150479/order-of-items-in-classes-fields-properties-constructors-methods

tiamato replied:
Исправлено

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
 
-                obj1x += random.Next(-1, 1);
-                obj1y += random.Next(-1, 1);
+    public class StepRange
```
Бесмысленный класс. То что тебе нужно для движения - это нормализованный вектор

tiamato replied:
переименовал в StepRangeRestrictions, чтобы было понятно что это

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
 
-                if (obj3x < 0)
-                    obj3x = 0;
+        private int CompareTo(GameObject gameObject)
```
Это ошибка определять этот метод от координат

tiamato replied:
Вместо него сделал статический компаратор позиции для использования в Sort()

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
+
+        public IEnumerable<GameObject> GetAliveItems()
+        {
+            return _items.Where(item => item.IsAlive);
```
Мёртвые объекты должны удаляться со сцены тем самым мы избавим объект от галочки IsAlive

tiamato replied:
Нельзя, ибо изменится поведение системы: в первоначальной реализации столкновение с "мертвым" объектом тоже приводит к "смерти" объекта.

Для решения необходимо поправить постановку задачи или первоначальную реализацию

tiamato replied:
По указанию дизайнера системы "смерть" от столкновения с "мертвым" объектом считается ошибкой реализации - исправлено

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -1,89 +1,229 @@
+        }
+    }
+
+    public class SceneController
```
Класс излишний, ну или это не контроллер вовсе. Это скорей SceneLoop

tiamato replied:
Переименовал

Commit:
Реализация позиции объекта с помощью Vector2

Commit:
Исправлен порядок следования инструкций в коде

Commit:
Исправление ошибок оформлению кода

Commit:
Игровые объекты при столкновении удаляются со сцены

Commit:
переименован класс StepRangeHolder -> StepRangeRestrictions

Commit:
Переименован класс StepRangeHolder -> StepRangeRestrictions

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,234 @@
+        public int X { get; }
+        public int Y { get; }
+
+        public int CompareTo(Vector2 vector)
```
Один вектор не может быть больше или меньше другого (на самом деле может если мы говорим про длину вектора) по этому CompareTo определять не корректно

tiamato replied:
определил метод Equals (условие равенства векторов: вектора равны, если их координаты равны)

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,234 @@
+
+        private void KillGameObjectsWithCollision()
+        {
+            _items.Sort(GameObject.PositionComparison);
```
Сортировка слишком дорогая операция для такой задачи

tiamato replied:
Убрал

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,234 @@
 
-                if (obj3x < 0)
-                    obj3x = 0;
+        public static int PositionComparison(GameObject x, GameObject y)
```
Position.Comparsion

tiamato replied:
Убрал, метод больше не используется

Commit:
Изменил алгоритм вычисления столкновений

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,208 @@
+        }
+    }
+
+    public class Vector2
```
Это лучше смоделировать структурой

tiamato replied:
сделал структурой

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,208 @@
-                    isalive2 = false;
-                    isalive3 = false;
-                }
+    public class StepRangeRestrictions
```
Всё ещё плохое название

tiamato replied:
убрал, см. ниже

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,208 @@
+        }
+    }
+
+    public class ConsoleView
```
В этом классе нет смысла

tiamato replied:
убрал

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,208 @@
+            OnSceneUpdated?.Invoke();
+        }
+
+        private void KillGameObjectsWithCollision()
```
Запутанные пляски с i-- и j--

tiamato replied:
упростил, без оптимизации

В этом месте пришлось менять алгоритм по сравнению с первоначальным, так как объекты физически удаляются из сцены и обрабатывается вариант одновременного столкновения всех объектов

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,208 @@
-                }
+    public class StepRangeRestrictions
+    {
+        public readonly int MinX;
+        public readonly int MaxX;
+        public readonly int MinY;
+        public readonly int MaxY;
```
Смысл этого класс мне не понятен. Если хочешь формализовать лимиты для оси то у тебя дубляж (две пары одна для X другая для Y хотя пара - это тип, а пара для X\Y - это объект этого типа) 

При работе с движением также есть другой аппарат: 
1) Ты получаешь случайной НОРМАЛИЗОВАННЫЙ вектор направления 
2) Умножаешь его на скорость 

Т.е клэмпить -1.0.1 - нет смысл так, как никакой конфигурации внешней у нас не предусмотрено

tiamato replied:
Это-то все понятно, но у нас нет задачи переписать _алгоритм_ работы системы.
Предполагал, что это лимиты шага изменения координат объекта управления (GameObject), потому, что движение объектов в системе реализовано именно так.
Так как для всех объектов в исходной реализации параметры движения одинаковые, делаем вывод, что это ограничения системы, то есть система не расширяется в части добавления объектов с параметрами движения, отличными от существующих или изменения этих параметров в существующих объектах

Убрал
Требования постановки задачи соблюдены

Commit:
Удалены классы StepRangeRestriction, ConsoleView
Vector2 теперь структура
упрощен код проверок

Commit:
Исправлены ошибки форматирования

Commit:
Параметры изменения шага GameObject вынесены в константы

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,127 @@
+
+        private static int RandomStep(int value, Random random)
+        {
+            // Интервал изменения [min, max); 
```
Комментарии в коде - это дурная привычка.

tiamato replied:
спорно, но в данном случае убрал

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,127 @@
+
+        public void AddGameObject(GameObject gameObject) => _items.Add(gameObject);
+
+        public void UpdateScene()
```
Просто Update

tiamato replied:
исправлено

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,127 @@
-                    obj1y = 0;
+    public class GameObject
+    {
+        private const int MinBoundaryRandomStepInclusive = -1;
+        private const int MaxBoundaryRandomStepExclusive = 1;
```
Настолько длинное название, что его лень читать и оно предстаёт скорей визуальным мусором нежели смысловой единицей.

Я не думаю что -1 и 1 будет являться магическим числом и это обязательно нужно оформить как константу. 
Но если прям очень хочется то попробуй названия Left и Right

tiamato replied:
убрал

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,127 @@
+            Console.Write(_name);
+        }
+
+        private static int RandomStep(int value, Random random)
```
Сделай метод расширения для Random 

И у тебя получится типа такого: 

position.x += random.GetDirection(); 

Или даже двухосевую 

position += random.GetDirection();

tiamato replied:
сделал вторым способом
одна особеннjсть - надо помнить что присвоение позиции должно делаться через Position, ибо там контроль ограничений

Commit:
Движение реализовано с использованием метода расширения Random для генерации шага

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,116 @@
+        private Vector2 Position
+        {
+            get => _position;
+            set => _position = new Vector2(value.X < 0 ? 0 : value.X, value.Y < 0 ? 0 : value.Y);
```
Нет в сетере ни в коем случае такие кренделя выписывать ненужно

tiamato replied:
вынес в класс Position

HolyMonkey commented on medium8-1-2.cs:
```diff
@@ -3,87 +3,116 @@
+            PrintAliveGameObjects();
+        }
+
+        private void PrintAliveGameObjects()
```
Можно просто Print

tiamato replied:
поправил

Commit:
Добавил класс, реализующий позицию с ограничениями
